thetis/callback.py

```diff
@@ -481,3 +481,43 @@ def __call__(self):
             field_vals.append(np.reshape(self._evaluate_field(field_name), (ndetectors, -1)))
 
         return np.hstack(field_vals)
+
+
+class FunctionalCallback(DiagnosticCallback):
+    """Callback that evaluates a functional involving integrals in both time and space. This callback can also be used
+    to assemble time dependent objective functionals for adjoint simulations. Time integration is achieved using the
+    trapezium rule."""
+    variable_names = ['current integral', 'objective value']
+
+    def __init__(self, scalar_callback, solver_obj, **kwargs):
+        """
+        :arg scalar_callback: Python function that returns a list of values of an objective functional.
+        :arg solver_obj: Thetis solver object
+        :arg **kwargs: any additional keyword arguments, see DiagnosticCallback
+        """
+        kwargs.setdefault('export_to_hdf5', False)
+        kwargs.setdefault('append_to_log', False)
+        super(FunctionalCallback, self).__init__(solver_obj, **kwargs)
+        self.scalar_callback = scalar_callback      # Evaluate functional
+        self.objective_value = []   # Store functional value
+        self.dt = solver_obj.options.timestep
+
+    def __call__(self):
+        value = self.scalar_callback()
+        self.objective_value.append(value)
+
+        return value
+
+    def get_vals(self):
+        return self.objective_value
+
+    def message_str(self, *args):
+        line = '{0:s} value {1:11.4e}'.format(self.name, args[0])
+        return line
+
+    def quadrature(self):
+        func = self.objective_value
+        J = 0
+        for i in range(1, len(func)):
+            J += 0.5 * (func[i] + func[i-1]) * self.dt
```

Comment:
You could do the update in a better way. For example, store a temporary integrant value in the class, and just increment it every time:
```python
self.integrant = 0
self.old_value = None

scalar_value = self.scalar_callback()
if self.old_value is not None:
    self.integrant += 0.5 * (self.old_value + scalar_value) * self.dt
self.old_value = scalar_value
```

Then you do not need to store all the values in the `self.objective_value` list at all.

Reply:
Thanks very much for the suggestions. The reason I was storing the spatial integrals at each timestep is so that I could look at their time evolution afterwards. But you are right, for more general purposes, it would be better not to store them. A user could just use a custom callback if this is something they want to do.

Reply:
Also I suppose those values can be retrieved from the log file anyway.